### DISCUSSION TITLE
Hardware updates while debugging

### DISCUSSION THREAD

**Discussion by cfwdman** [dirkwhoffmann/virtualc64#771] [Ideas]
In an nostalgic moment some weeks ago I rediscovered my old interest in c64 programming... and found there's an active scene using various tools for assembler programming and debugging. I currently use visual studio code with the "Kick Ass (C64)" / KickAssembler to generate my prg files... 

As I'm returning to assembler after a looong time I needed and wanted to debug the code quite extensively and I came to observe an issue which I'd like to share and discuss: Debugging appears to me to be possible in two ways: 
1. emulator supported debugging by placing breakpoints in the inspector or 
2. monitor based debugging by using a brk opcode and breaking into a debugger/monitor offered by e.g. "a final cartridge III"

Option 2 allows me to observe any changes in hardware registers both on the memory and on the screen/sound side of things: When I change he background color or write to the screen memory before the BRK is hit it's visible in the debugger. But placing/removing BRK opcodes in my code is tedious work, even if one just switches from BRK to NOP... 

Option 1 - the emu debugger in the inspector - is much more convenient by placing, deactivating, deleting breakpoints much more dynamically. BUT the caveat to me seems that updates to hardware are NOT propagated from memory to the screen display/sound... while using "step over" or the "run"*) button from break point to breakpoint as the whole emulator and not the processor alone seem "on hold" until the code is continued... which is indicated by the screen being dimmed.... and actually the inspector tab "VICII" showing that the counters are stopped when at a breakpoint... 

Am I missing an option? 
and Is my following interpretation correct? 
In the emulator debugger the "clock" for the whole machine seems to be stopped when hitting a breakpoint set by the inspector as while debugging natively with an BRK optcode The machine/EMU is still humming on (the vics?) clock while the processor just stops fetching, parsing and executing instructions based on the BRK opcode?

I think I understand from some reading that the 6510 processor does support breaking only by means of the brk opcode and not by other "external" breakpoints but I wonder if that could be enhanced? 

What would it take to use the emulators breakpoints/debugging tools and have the hardware clocking away? 
Looking forward to you - probably much more - educated opinion... 

Happy emulating!

*) on a side note: I realised that in a test with a loop writing to the screen mem the look is cycled through multiple times (2-4 times) after hitting the run button by "running over the breakpoint" multiple times... 

BTW: VICE and Retro Debugger seem to be working the same way... 
no hardware/screen updates which stepping trough code / running from breakpoint to breakpoint

please find the test code below - with the place of the breakpoint: 
`

        // Start the basic program
        *=$0800 "BASIC Start"  
        // location to put a 1 line basic program so we can just
        // type run to execute the assembled program.
        
        // will just call assembled program at correct location
        //    10 SYS (4096)

        // These bytes are a one line basic program that will 
        // do a sys call to assembly language portion of
        // of the program which will be at $1000 or 4096 decimal
        // basic line is: 
        // 10 SYS (4096)
        .byte $00                // first byte of basic should be a zero
        .byte $0E, $08           // Forward address to next basic line
        .byte $0A, $00           // this will be line 10 ($0A)
        .byte $9E                // basic token for SYS
        .byte $20, $28, $34, $30, $39, $36, $29 // ASCII for " (4096)"
        .byte $00, $00, $00      // end of basic program (addr $080E from above)

        // assembler constants for special memory locations
        .const CLEAR_SCREEN_KERNAL_ADDR = $E544     // Kernal routine to clear screen
        .const PRINT_STRING_BASIC_ADDR = $AB1E      // Basic routine to print text
        .const SCREEN_START = $0400                 // The start of c64 screen memory

        // a somewhat random location in screen memory to write to directly
        .const SCREEN_DIRECT_START = SCREEN_START + 10 * 40 + (40 - 18) / 2

        // program variables
        str_to_print: .text @"HELLO VIA BASIC-KERNAL\$00"  // null terminated string to print
                                                // via the BASIC routine

        str_to_poke: .text  @"hello direct poked\$00"  // null terminated string to print
                                                // via copy direct to screen memory

        // our assembly code will goto this address
        *=$1000 "Main Start"

        // clear screeen leave cursor upper left
        jsr CLEAR_SCREEN_KERNAL_ADDR 
        
        // method 1 call basic routine since we cleared screen 
        // above the string will start in upper left
        lda #<str_to_print           // LSB of addr of string to print to A
        ldy #>str_to_print           // MSB of addr of str to print to Y
        jsr PRINT_STRING_BASIC_ADDR  // call kernal routine to print the string

        // method 2 write direct to screen memory
        ldx #0                  // use x reg as loop index start at 0
        DirectLoop:
        lda str_to_poke,x         // put a byte from string into accum
        beq Done                // if the byte was 0 then we're done 
        sta SCREEN_DIRECT_START,x // Store the byte to screen
        // brk
        inx                     // inc to next byte and next screen location 
        jmp DirectLoop          // Go back for next byte
        Done:      
        rts                     // program done, return
`

**Comment by dirkwhoffmann** [comment]
I'll try to clarify what’s going on. The emulator updates the displayed texture each time a frame is completed. Therefore, you would observe a screen update when the 'Step into' button is pressed repeatedly until the first CPU cycle of a new frame is reached. The only distinction from a real C64 is that the emulator updates the screen periodically in one chunk at 50Hz or 60Hz, depending on the VIC revision (PAL/NTSC), while the cathode ray of a CRT draws the screen continuously. 

BTW, there is no substantial technical difference between 'single-step mode' and 'continuous mode' from the emulator's perspective. If 'Step Into' is pressed, the emulator starts running and automatically returns to pause mode after a single cycle has been executed, whereas in continuous mode, it automatically pauses for a little while after a single frame has been completed.

**Reply by cfwdman** [reply]
Thank you for the interesting clarification. 
 
Would it be possible to allow for the dynamic breakpoints set by the emulator to cause the same effects in the processors emulation code as the execution of the BRK opcode? This could be an alternative option to be set in the preferences/inspector to the current "time freeze" of the whole system. This would allow then to break into a native debugger/monitor (e.g. from a cartridge)... 
 
Much more complex and rather "nice to have" it would be great to allow for the emulators monitor/inspector to be triggered by tweaking the handling of the interrupt vector locations (or interpretation of the content thereof). 

With respect to the step/run feature: 
- Why can I observe that a single click on the run command steps multiple times over a breakpoint?  
- With "single cycle" I assume you mean processor cycle (fully working on an opcode) rather than a "clock cycle"?

**Comment by dirkwhoffmann** [comment]
> Would it be possible to allow for the dynamic breakpoints set by the emulator to cause the same effects in the processors emulation code as the execution of the BRK opcode? 

Yes, I think this should be doable. I've opened a new issue for this (#773).

> Why can I observe that a single click on the run command steps multiple times over a breakpoint?

This shouldn't be the case. Sounds like a bug.

> With "single cycle" I assume you mean processor cycle (fully working on an opcode) rather than a "clock cycle"?

Yes, you're right. The proper term should be *instruction* not *clock cycle*.
